modules/eks_managed_nodes/asg_tags.tf

```diff
@@ -0,0 +1,71 @@
+locals {
+  taint_effects = {
+    NO_SCHEDULE        = "NoSchedule"
+    NO_EXECUTE         = "NoExecute"
+    PREFER_NO_SCHEDULE = "PreferNoSchedule"
+  }
+
+  cluster_autoscaler_label_tags = {
+    for name, group in local.eks_managed_node_groups : name => merge(
+      {
+        for label_name, label_value in group.labels : "k8s.io/cluster-autoscaler/node-template/label/${label_name}" => label_value
+      },
+    ) if try(group.labels, {}) != {}
+  }
+
+  cluster_autoscaler_taint_tags = {
+    for name, group in local.eks_managed_node_groups : name => merge(
+      {
+        for taint in group.taints : "k8s.io/cluster-autoscaler/node-template/taint/${taint.key}" => "${taint.value}:${local.taint_effects[taint.effect]}"
+      },
+    ) if try(group.taints, {}) != {}
+  }
+
+  cluster_autoscaler_implicit_tags = {
+    for name, group in local.eks_managed_node_groups : name => merge(
+      length(try(group.instance_types, local.eks_managed_node_group_defaults.instance_types)) == 1 ? {
+        "k8s.io/cluster-autoscaler/node-template/label/node.kubernetes.io/instance-type" = one(try(group.instance_types, local.eks_managed_node_group_defaults.instance_types))
+      } : {},
+      length(data.aws_autoscaling_group.node_groups[name].availability_zones) == 1 ? {
+        "k8s.io/cluster-autoscaler/node-template/label/topology.ebs.csi.aws.com/zone" = one(data.aws_autoscaling_group.node_groups[name].availability_zones)
+      } : {},
+      length(data.aws_autoscaling_group.node_groups[name].availability_zones) == 1 ? {
+        "k8s.io/cluster-autoscaler/node-template/label/topology.kubernetes.io/zone" = one(data.aws_autoscaling_group.node_groups[name].availability_zones)
+      } : {},
+    )
+  }
+
+  cluster_autoscaler_asg_tags = {
+    for name, group in local.eks_managed_node_groups : name => merge(
+      try(local.cluster_autoscaler_label_tags[name], {}),
+      try(local.cluster_autoscaler_taint_tags[name], {}),
+      local.cluster_autoscaler_implicit_tags[name],
+    )
+  }
+}
+
```

Comment:
Minor suggestion to try and reduce the use of `try` in `local.cluster_autoscaler_asg_tags`.

In `local.cluster_autoscaler_label_tags` and `local.cluster_autoscaler_taint_tags`, you can remove the `if` clause in the `for` expression to always produce a value for each group. (And remove the `merge`)

```hcl
  cluster_autoscaler_label_tags = {
    for name, group in local.eks_managed_node_groups : name => {
        for label_name, label_value in try(group.labels, {}) : "k8s.io/cluster-autoscaler/node-template/label/${label_name}" => label_value
    }
  }
```

Might be a syntax error for the above in terms of line breaks. Need to tweak the formatting a bit.

Reply:
Thank you @lawliet89 . Done.